### PR TITLE
Refactor netconf logging to create log file per device

### PR DESCRIPTION
**Pull request by dwapstra**
Updated the yang.connector.Netconf implementation to create a log file for each connection instance.

### CODE REVIEW

**Comment by dwapstra** [comment]
@miott  please review this change where the netconf connection will create a log file for each connection instance

File: connector/src/yang/connector/netconf.py [hunk 1]
```diff
@@ -30,6 +32,88 @@ class BaseConnection:
...
+    MAX_LINES = 40
+
+    def format(self, record):
+        msg = super().format(record)
+        lines = msg.splitlines()
+        msg_len = len(lines)
+        if msg_len > self.MAX_LINES:
+            return '\n'.join(lines[0:20] + [f'\n\n... skipping {msg_len-40} lines ...\n\n'] + lines[-20:])
+        else:
+            return '\n'.join(lines)
```
Comment (miott) [inline connector/src/yang/connector/netconf.py:97]: Why split NETCONF messages?  The rpc-reply could be large and users may want to copy/paste it from log.

File: connector/src/yang/connector/netconf.py [hunk 2]
```diff
@@ -383,12 +524,15 @@ def connect(self):
...
             self.session.connect(**defaults)
-            logger.info('NETCONF CONNECTED')
+            self.log.info('NETCONF CONNECTED')
         except Exception:
             if self.session.transport:
                 self.session.close()
             raise
 
+        # disable info logging for ncclient
+        nccl.setLevel(logging.WARNING)
```
Comment (miott) [inline connector/src/yang/connector/netconf.py:534]: What messages are we loosing by setting this to WARNING?  Users should see close to what ncclient is seeing over the wire in order to debug device issues.

File: connector/src/yang/connector/netconf.py (hunk 1 above)
Comment (dwapstra) [reply connector/src/yang/connector/netconf.py:97]: The response can be quite large, this is only done for console / tasklog messages. The full response is availble in the per device/connection log file (which is part of this change)

File: connector/src/yang/connector/netconf.py (hunk 2 above)
Comment (dwapstra) [reply connector/src/yang/connector/netconf.py]: I have updated the logging level change to only change it right before we call the low level request. The device log will contain the sent and received XML messages, by default they are formatted in pretty-print form.

File: connector/src/yang/connector/netconf.py (hunk 1 above)
Comment (miott) [reply connector/src/yang/connector/netconf.py:97]: Not really clear about what a ``console`` or ``tasklog`` message is.  What I do know about running pyats tests is that I have log messages show up in my terminal when running tests, and, I also have the same log messages showing up in the etrade link in my web browser.  I have used both of those sources to copy/paste content into my VS editor to examine or reuse messages.  With this change, I would have to find and open the log file.  Were there a multiple requests from users to shorten these messages?  If not, can we leave them as-is?

File: connector/src/yang/connector/netconf.py (hunk 2 above)
Comment (miott) [reply connector/src/yang/connector/netconf.py]: ncclient logging has changed over the years.  As I remember it, the sent RPC used to show up in INFO logs but the rpc-reply showed up in DEBUG level.  That was probably to protect user content from being added to logs unless the user changed to DEBUG level.  If the rpc-reply now shows up at INFO level, this change is fine.

**Review by miott — APPROVED** [review]

**Review by ademz — APPROVED** [review]

File: connector/src/yang/connector/netconf.py (hunk 1 above)
Comment (dwapstra) [reply connector/src/yang/connector/netconf.py:97]: Will add some more enhancements for the user control of log messages before merging this

**Comment by oboehmer** [comment]
@dwapstra , when do you think this can be merged and released?

**Comment by Taarini** [comment]
Sorry about the delay. This was a draft PR so I changed to ready for review. Now it can be merged

**Comment by Taarini** [comment]
Can you please merge the latest dev to your branch and then you are ready for merge

File: connector/src/yang/connector/netconf.py (hunk 1 above)
Comment (miott) [reply connector/src/yang/connector/netconf.py:97]: > Not really clear about what a `console` or `tasklog` message is. What I do know about running pyats tests is that I have log messages show up in my terminal when running tests, and, I also have the same log messages showing up in the etrade link in my web browser. I have used both of those sources to copy/paste content into my VS editor to examine or reuse messages. With this change, I would have to find and open the log file. Were there a multiple requests from users to shorten these messages? If not, can we leave them as-is?

``MAX_LINES`` is still hard-coded to 40 for RPC reporting.  If you want to keep it this way we can do so and modify later if users complain.

**Comment by oboehmer** [comment]
@dwapstra , any chance you can merge this and create a release? Would rather not wait until 23.1 at the end of January..

**Comment by dwapstra** [comment]
Example output with default settings. Full connection log is stored in the netconf log file.

```
2023-01-12T13:07:19: %YANG-INFO: +++ ott-c9300-19 netconf logfile /Users/dwapstra/p39/users/dwapstra/runinfo/ncTest_job.2023Jan12_13:07:13.923008/ott-c9300-19-nc-20230112T130719244.log +++

2023-01-12T13:07:32: %UNICON-INFO: Device 'ott-c9300-19' connection 'netconf' via new SSH tunnel 127.0.0.1:20004
2023-01-12T13:07:32: %NCCLIENT-INFO: Connected (version 2.0, client OpenSSH_7.9)
2023-01-12T13:07:34: %NCCLIENT-INFO: Authentication (password) successful!
2023-01-12T13:07:40: %NCCLIENT-INFO: [host 127.0.0.1 session 0x1125990a0] Sending:
2023-01-12T13:07:40: %NCCLIENT-INFO: <nc:hello xmlns:nc="urn:ietf:params:xml:ns:netconf:base:1.0">
2023-01-12T13:07:40: %NCCLIENT-INFO:   <nc:capabilities>
2023-01-12T13:07:40: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:base:1.0</nc:capability>
2023-01-12T13:07:40: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:base:1.1</nc:capability>
2023-01-12T13:07:40: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:writable-running:1.0</nc:capability>
2023-01-12T13:07:40: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:candidate:1.0</nc:capability>
2023-01-12T13:07:40: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:confirmed-commit:1.0</nc:capability>
2023-01-12T13:07:40: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:rollback-on-error:1.0</nc:capability>
2023-01-12T13:07:40: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:startup:1.0</nc:capability>
2023-01-12T13:07:40: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:url:1.0?scheme=http,ftp,file,https,sftp</nc:capability>
2023-01-12T13:07:40: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:validate:1.0</nc:capability>
2023-01-12T13:07:40: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:xpath:1.0</nc:capability>
2023-01-12T13:07:40: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:notification:1.0</nc:capability>
2023-01-12T13:07:40: %NCCLIENT-INFO:     <nc:capability>urn:liberouter:params:netconf:capability:power-control:1.0</nc:capability>
2023-01-12T13:07:40: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:interleave:1.0</nc:capability>
2023-01-12T13:07:40: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:with-defaults:1.0</nc:capability>
2023-01-12T13:07:40: %NCCLIENT-INFO:   </nc:capabilities>
2023-01-12T13:07:40: %NCCLIENT-INFO: </nc:hello>
2023-01-12T13:07:40: %NCCLIENT-INFO: ]]>]]>
2023-01-12T13:07:41: %NCCLIENT-INFO: [host 127.0.0.1 session 0x1125990a0] Received message from host
2023-01-12T13:07:41: %NCCLIENT-INFO: [host 127.0.0.1 session-id 35] initialized: session-id=35 | server_capabilities=<dict_keyiterator object at 0x112598180>
2023-01-12T13:07:41: %NETCONF-INFO: NETCONF CONNECTED
2023-01-12T13:07:41: %NETCONF-INFO: Sending rpc...
2023-01-12T13:07:41: %NETCONF-INFO: <rpc xmlns="urn:ietf:params:xml:ns:netconf:base:1.0" message-id="101">
  <get-config>
    <source>
      <running/>
    </source>
  </get-config>
</rpc>
2023-01-12T13:07:57: %NETCONF-INFO: Receiving rpc-reply after 16.301 sec...
2023-01-12T13:07:57: %NETCONF-INFO: <rpc-reply xmlns="urn:ietf:params:xml:ns:netconf:base:1.0" message-id="101">
  <data>
    <cloud-services-cfg-data xmlns="http://cisco.com/ns/yang/Cisco-IOS-XE-cloud-services-cfg">
      <cloudm-config/>
    </cloud-services-cfg-data>
    <native xmlns="http://cisco.com/ns/yang/Cisco-IOS-XE-native">
      <version>17.12</version>
      <memory>
        <free>
          <low-watermark>
            <processor>131078</processor>
          </low-watermark>
        </free>
      </memory>
      <service>
        <timestamps>
          <debug-config>
            <datetime>
              <msec/>
            </datetime>


... skipping 5624 lines ...


                <role xmlns:oc-aaa-types="http://openconfig.net/yang/aaa/types">oc-aaa-types:SYSTEM_ROLE_ADMIN</role>
              </config>
            </user>
          </users>
        </authentication>
        <authorization>
          <events>
            <event>
              <event-type xmlns:oc-aaa-types="http://openconfig.net/yang/aaa/types">oc-aaa-types:AAA_AUTHORIZATION_EVENT_CONFIG</event-type>
              <config>
                <event-type xmlns:oc-aaa-types="http://openconfig.net/yang/aaa/types">oc-aaa-types:AAA_AUTHORIZATION_EVENT_CONFIG</event-type>
              </config>
            </event>
          </events>
        </authorization>
      </aaa>
    </system>
  </data>
</rpc-reply>
2023-01-12T13:07:57: %AETEST-INFO: The result of section test is => PASSED

```

**Pull request merged**